File: x3270/ConvTools/Edbcs_ucm.pl
#!/usr/bin/perl
# Grab the DBCS part of a DBCS or MBCS ucm file and output DBCS translation
# tables.

my $file = $ARGV[0];
my $in = 0;
my %d2u;
my %u2d;
open TABLE, "<". $file . ".ucm" or die "No table.";
while (<TABLE>) {
    chomp;
    next if /^#/;
    last if /^END CHARMAP/;
    if (/^CHARMAP/) {
	$in = 1;
	next;
    }
    next if (! $in);
    # <U0020> \x20 #|0 SPACE
    if (/<U([0-9a-fA-F]+)> +\\x([0-9a-fA-F]+)\\x([0-9a-fA-F]+)/) {
	my $u = hex($1);
	my $d = hex($2 . $3);
	#printf "got U+%04x -> X'%02X'\n", $u, $d;
	$d2u{$d} = $u;
	$u2d{$u} = $d;
    } elsif (/<U([0-9a-fA-F]+)> +\\x([0-9a-fA-F]+)/) {
    } else {
	die "Eh?";
    }
}

print "/* Unicode to EBCDIC DBCS translation table for $file */\n";
my $row = "";
my $i;
my $any = 0;
foreach $i (0..65535) {
    if ($i && !($i % 128)) {
	printf "/* %04x */ ", $i - 128;
	if ($any) {
	    print "\"$row\",\n";
	} else {
	    print "NULL,\n";
	}
	$row = "";
	$any = 0;
    }
    if ($u2d{$i}) {
	$row .= sprintf("\\x%02x\\x%02x",
	    ($u2d{$i} >> 8) & 0xff, $u2d{$i} & 0xff);
	$any = 1;
    } else {
	$row .= "\\x00\\x00";
    }
}
print "/* ff80 */ ";
if ($any) {
    print "\"$row\"";
} else {
    print "NULL";
}
print " };\n";

print "/* EBCDIC DBCS to Unicode translation table for $file */\n";
$row = "";
$any = 0;
foreach $i (0..65535) {
    if ($i && !($i % 128)) {
	printf "/* %04x */ ", $i - 128;
	if ($any) {
	    print "\"$row\",\n";
	} else {
	    print "NULL,\n";
	}
	$row = "";
	$any = 0;
    }
    if ($d2u{$i}) {
	$row .= sprintf("\\x%02x\\x%02x",
	    ($d2u{$i} >> 8) & 0xff, $d2u{$i} & 0xff);
	$any = 1;
    } else {
	$row .= "\\x00\\x00";
    }
}
print "/* ff80 */ ";
if ($any) {
    print "\"$row\"";
} else {
    print "NULL";
}
print " };\n";

File: ConvTools/Edbcs_ucm.py
# Translate a .ucm file to data structures for unicode_dbcs.c.

import os
import re
import sys
from typing import Dict

class dbcs_ucm:
    def __init__(self, file: str):
        self.filename = file
        self.cs = os.path.splitext(os.path.basename(self.filename))[0]
        self.cp = m.group(1) if (m := re.match(r'^ibm-([0-9]+)_.*', self.cs)) else "???"
        self.d2u = {}
        self.u2d = {}
    
    def read(self):
        '''Read the .ucm file'''
        dmatch = re.compile(r'^<U([0-9a-zA-Z]+)> +\\x([0-9a-zA-Z]+)\\x([0-9a-zA-Z]+).*')
        inside = False
        with open(self.filename) as f:
            for line in f:
                if line.startswith('#'):
                    continue
                if line.startswith('END CHARMAP'):
                    break
                if line.startswith('CHARMAP'):
                    inside = True
                    continue
                if not inside:
                    continue
                if m := dmatch.match(line):
                    u = int(m.group(1), 16)
                    d = int(m.group(2) + m.group(3), 16)
                    self.d2u[d] = u
                    self.u2d[u] = d

    def dump1(self, text: str, dict: Dict[int, int], end = ''):
        '''Write out one of the two tables'''
        print(f'/* {text} translation table for {self.cs} */ ' + '{')
        row = ''
        any = False
        for i in range(65536):
            if i > 0 and (i % 128) == 0:
                print(f'/* {i-128:04x} */ ', end='')
                if any:
                    print(f'"{row}",')
                else:
                    print('NULL,')
                row = ''
                any = False
            if i in dict:
                row += '\\x{0:02x}\\x{1:02x}'.format((dict[i] >> 8) & 0xff, dict[i] & 0xff)
                any = True
            else:
                row += '\\x00\\x00'
        print('/* ff80 */ ', end='')
        if any:
            print(f'"{row}"', end='')
        else:
            print('NULL', end='')
        print(' }' + end)

    def dump(self):
        '''Write out the translation of the file'''

        print('    { ' + f'"cp{self.cp}", "<codepage>" /* yyy, zzz */,')

        # Unicode to EBCDIC.
        self.dump1('Unicode to EBCDIC DBCS', self.u2d, ',')

        # EBCDIC to Unicode.
        self.dump1('EBCDIC DBCS to Unicode', self.d2u)

        print('    },')

if len(sys.argv) != 2:
    print(f'Usage: {sys.argv[0]} ucm-file', file=sys.stderr)
    exit(1)
d = dbcs_ucm(sys.argv[1])
d.read()
d.dump()

